src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+                            }
+                            stack.pop()?;
+
+                            stack.push_back(if success {
+                                vch_true.clone()
+                            } else {
+                                vch_false.clone()
+                            });
+
+                            if op == OP_CHECKMULTISIGVERIFY {
+                                if success {
+                                    stack.pop()?;
+                                } else {
+                                    return set_error(ScriptError::CheckMultisigVerify);
+                                }
+                            }
+                        }
+
+                        _ => {
+                            return set_error(ScriptError::BadOpcode);
+                        }
+                    }
+                }
+            }
+        }
+
+        // Size limits
+        if stack.size() + altstack.size() > 1000 {
+            return set_error(ScriptError::StackSize);
+        }
+    }
+
+    if !vexec.empty() {
+        return set_error(ScriptError::UnbalancedConditional);
+    }
+
+    set_success(true)
+}
+
+/// All signature hashes are 32 bytes, since they are either:
+/// - a SHA-256 output (for v1 or v2 transactions).
+/// - a BLAKE2b-256 output (for v3 and above transactions).
+pub const SIGHASH_SIZE: usize = 32;
+
+/// A function which is called to obtain the sighash.
+///    - script_code: the scriptCode being validated. Note that this not always
+///      matches script_sig, i.e. for P2SH.
+///    - hash_type: the hash type being used.
+///
+/// The `extern "C"` function that calls this doesn’t give much opportunity for rich failure
+/// reporting, but returning `None` indicates _some_ failure to produce the desired hash.
+pub type SighashCalculator<'a> = &'a dyn Fn(&[u8], HashType) -> Option<[u8; SIGHASH_SIZE]>;
+
+impl CallbackTransactionSignatureChecker<'_> {
+    pub fn verify_signature(vch_sig: &[u8], pubkey: &PubKey, sighash: &[u8; SIGHASH_SIZE]) -> bool {
+        pubkey.verify(sighash, vch_sig)
+    }
+}
+
+impl SignatureChecker for CallbackTransactionSignatureChecker<'_> {
+    fn check_sig(&self, vch_sig_in: &[u8], vch_pub_key: &[u8], script_code: &Script) -> bool {
+        let pubkey = PubKey(vch_pub_key);
+        if !pubkey.is_valid() {
+            return false;
+        };
+
+        // Hash type is one byte tacked on to the end of the signature
+        match vch_sig_in.split_last() {
+            None => false,
+            Some((hash_type, vch_sig)) => HashType::from_bits((*hash_type).into(), false)
+                .ok()
+                .and_then(|hash_type| (self.sighash)(script_code.0, hash_type))
+                .map(|sighash| Self::verify_signature(vch_sig, &pubkey, &sighash))
+                .unwrap_or(false),
+        }
+    }
+
+    fn check_lock_time(&self, lock_time: &ScriptNum) -> bool {
+        // There are two times of nLockTime: lock-by-blockheight
```

Comment:
```suggestion
        // There are two kinds of nLockTime: lock-by-blockheight
```
Bitcoin fixed this typo in https://github.com/bitcoin/bitcoin/commit/e846b2a1e96f27257cbb79f7f61bfaf3d873bb97